ydb/docs/en/core/development/localdb-uncommitted-txs.md

```diff
@@ -0,0 +1,178 @@
+| 46 | 10 | .. column C (32-bit) |
+| 50 | 15 | .. TDelta::TxId |
+| 58 | 0x61 | Main: HasHistory + IsVersioned + ERowOp::Upsert |
+| 59 | 0x00 | .. key column is not NULL |
+| 60 | K | .. key column (32-bit) |
+| 64 | 0x01 | .. column A = ECellOp::Set |
+| 65 | 1 | .. column A (32-bit) |
+| 69 | 0x01 | .. column B = ECellOp::Set |
+| 70 | 2 | .. column B (32-bit) |
+| 74 | 0x01 | .. column C = ECellOp::Set |
+| 75 | 3 | .. column C (32-bit) |
+| 79 | 3000 | .. RowVersion.Step |
+| 87 | 12 | .. RowVersion.TxId |
+| 95 | - | End of record |
+
+The HasHistory flag in the Main record shows that other two records are stored among history data with keys `(RowId, 2000, 11)` and `(RowId, 1000, 10)` respectively.
+
+When compacting iterator runs in a special mode that enumerates all deltas and all versions for each key. The compaction scan implementation (see [flat_ops_compact.h](https://github.com/ydb-platform/ydb/blob/main/ydb/core/tablet_flat/flat_ops_compact.h)) first aggregates all uncommitted deltas by their TxId in the same order (in case changes from different TxIds overlap their order may change arbitrarily, which is OK since such transactions are not supposed to both commit). After uncommitted deltas are aggregated, they a flushed to the resulting SST (see [flat_part_writer.h](https://github.com/ydb-platform/ydb/blob/main/ydb/core/tablet_flat/flat_part_writer.h) and [flat_page_writer.h](https://github.com/ydb-platform/ydb/blob/main/ydb/core/tablet_flat/flat_page_writer.h)), and committed row versions for the same key are enumerated, which are written in decreasing version order.
+
+When iterator positions to the first committed delta (i.e. the IsDelta record which has commit info in the transaction map), the commit info is used as the resulting row versions, with row state combined from all deltas below, including the first committed record from each LSM level participating in compaction. When positioning to the next version iterator skips delta with version at or below the last one and the process is repeated.
+
+You may notice that with a large number of compacted deltas for a key, when they are later committed with different versions, the process of generating committed records grows quadratically in the number of deltas. For this reason upper levels should control the number of deltas for each key and must not allow them to grow too large (deltas might be merged in the reverse order in the future to side step this problem). Other limitation is that uncommitted deltas for a given key need to all be in memory and on a single page, since each read must walk the entire delta chain and check whether each record is committed. In the future we may want to store deltas across multiple pages, but since they all need to be in memory anyway there is little reason to do so. Removing the requirement for all deltas to be in memory during reads is theoretically possible, but it requires storing them in a different form.
```

Comment:
```suggestion
With a large number of compacted deltas for a key, when they are later committed with different versions, the process of generating committed records grows quadratically in the number of deltas. For this reason upper levels should control the number of deltas for each key and must not allow them to grow too large (deltas might be merged in the reverse order in the future to side step this problem). Other limitation is that uncommitted deltas for a given key need to all be in memory and on a single page, since each read must walk the entire delta chain and check whether each record is committed. In the future we may want to store deltas across multiple pages, but since they all need to be in memory anyway there is little reason to do so. Removing the requirement for all deltas to be in memory during reads is theoretically possible, but it requires storing them in a different form.
```